connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestForwardingIntegrationTest.java

```diff
@@ -0,0 +1,228 @@
+        );
+        if (firstException.get() != null) {
+            throw new RuntimeException("Unable to cleanly close resources", firstException.get());
+        }
+    }
+
+    @Test
+    public void testRestForwardNoSsl() throws Exception {
+        testRestForwardToLeader(false, false);
+    }
+
+    @Test
+    public void testRestForwardSsl() throws Exception {
+        testRestForwardToLeader(true, true);
+    }
+
+    @Test
+    public void testRestForwardLeaderSsl() throws Exception {
+        testRestForwardToLeader(false, true);
+    }
+
+    @Test
+    public void testRestForwardFollowerSsl() throws Exception {
+        testRestForwardToLeader(true, false);
+    }
+
+    public void testRestForwardToLeader(boolean followerSsl, boolean leaderSsl) throws Exception {
+        DistributedConfig followerConfig = new DistributedConfig(baseWorkerProps(followerSsl));
+        DistributedConfig leaderConfig = new DistributedConfig(baseWorkerProps(leaderSsl));
+
+        // Follower worker setup
+        followerClient = new RestClient(followerConfig);
+        followerServer = new RestServer(followerConfig, followerClient);
+        followerServer.initializeServer();
+        when(followerHerder.plugins()).thenReturn(plugins);
+        followerServer.initializeResources(followerHerder);
+
+        // Leader worker setup
+        leaderClient = new RestClient(leaderConfig);
+        leaderServer = new RestServer(leaderConfig, leaderClient);
+        leaderServer.initializeServer();
+        when(leaderHerder.plugins()).thenReturn(plugins);
+        leaderServer.initializeResources(leaderHerder);
+
+        // External client setup
+        factory = SSLUtils.createClientSideSslContextFactory(followerConfig);
+        factory.start();
+        SSLContext ssl = factory.getSslContext();
+        httpClient = HttpClients.custom()
+                .setSSLContext(ssl)
+                .build();
+
+        // Follower will forward to the leader
+        URI leaderUrl = leaderServer.advertisedUrl();
+        RequestTargetException forwardException = new NotLeaderException("Not leader", leaderUrl.toString());
+        ArgumentCaptor<Callback<Herder.Created<ConnectorInfo>>> followerCallbackCaptor = ArgumentCaptor.forClass(Callback.class);
+        doAnswer(invocation -> {
+            followerCallbackCaptor.getValue().onCompletion(forwardException, null);
+            return null;
+        }).when(followerHerder)
+                .putConnectorConfig(any(), any(), anyBoolean(), followerCallbackCaptor.capture());
+
+        // Leader will reply
+        Herder.Created<ConnectorInfo> leaderAnswer = new Herder.Created<>(true, null);
+        ArgumentCaptor<Callback<Herder.Created<ConnectorInfo>>> leaderCallbackCaptor = ArgumentCaptor.forClass(Callback.class);
+        doAnswer(invocation -> {
+            leaderCallbackCaptor.getValue().onCompletion(null, leaderAnswer);
+            return null;
+        }).when(followerHerder)
```

Comment:
Should this be the leader?
```suggestion
        }).when(leaderHerder)
```